### ISSUE TITLE
feature request: jest.mock - require factory function

### ISSUE THREAD

**Issue by dylang** [jest-community/eslint-plugin-jest#741]
I loved the auto-mocking feature of `jest.mock` when learning Jest, but now with an enterprise-sized project, it can cause unexpected problems when modules have side effects on `import`.

_Example where `jest.mock` automocking can cause problems:_

```ts
// ./get-config.ts - file to be mocked
import mem from 'mem';

// mem is being executed on import
export const foo = mem(() => ({});
```

```ts
// ./server.test.ts - test file

jest.mock('mem');
jest.mock('./get-config'); // ERROR HERE because the mock mem is not returning a function.
```

# Rule fail

```ts
jest.mock('mem'); // no factory function
jest.mock('./get-config'); // no factory function
```

# Rule pass

```ts
jest.mock('mem', () => jest.fn());
jest.mock('mem', () => (fn: Function) => fn);
jest.mock('mem', () => jest.fn((fn: Function) => fn));
jest.mock('./get-config', () => ({ foo: jest.fn() }));
jest.mock('./get-config', () => jest.fn());
```

**Comment by G-Rath** [comment]
I use mocks in most of my projects, but just about never used the factory parameter without any issues - could you provide more detail on what problem this prevents?

Otherwise, I believe you should be able to craft a selector for [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax) that'd do this for you - you'll want to match on all `jest.mock` calls that have only one argument.

**Comment by G-Rath** [comment]
@SimenB do you have any thoughts on this? otherwise I might close it due to lack of activity since I can't see a clear problem.

(cc @dylang)

**Comment by G-Rath** [comment]
I think let's go with recommending `no-restricted-syntax` for this for now, given the lack of activity and I've not seen any "+1"ings on this issue or in others/comments.

I'm happy to add a new rule if a few more people request this.
